compiler/rustc_lint/src/let_underscore.rs

```diff
@@ -0,0 +1,270 @@
+    /// ```compile_fail
+    /// use std::sync::{Arc, Mutex};
+    /// use std::thread;
+    /// let data = Arc::new(Mutex::new(0));
+    ///
+    /// thread::spawn(move || {
+    ///     // The lock is immediately released instead of at the end of the
+    ///     // scope, which is probably not intended.
+    ///     let _ = data.lock().unwrap();
+    ///     println!("doing some work");
+    ///     let mut lock = data.lock().unwrap();
+    ///     *lock += 1;
+    /// });
+    /// ```
+    /// ### Explanation
+    ///
+    /// Statements which assign an expression to an underscore causes the
+    /// expression to immediately drop instead of extending the expression's
+    /// lifetime to the end of the scope. This is usually unintended,
+    /// especially for types like `MutexGuard`, which are typically used to
+    /// lock a mutex for the duration of an entire scope.
+    ///
+    /// If you want to extend the expression's lifetime to the end of the scope,
+    /// assign an underscore-prefixed name (such as `_foo`) to the expression.
+    /// If you do actually want to drop the expression immediately, then
+    /// calling `std::mem::drop` on the expression is clearer and helps convey
+    /// intent.
+    pub LET_UNDERSCORE_LOCK,
+    Deny,
+    "non-binding let on a synchronization lock"
+}
+
+declare_lint! {
+    /// The `let_underscore_must_use` lint checks for statements which don't bind
+    /// a `must_use` expression to anything, causing the lock to be released
+    /// immediately instead of at end of scope, which is typically incorrect.
+    ///
+    /// ### Example
+    /// ```rust
+    /// #[must_use]
+    /// struct SomeStruct;
+    ///
+    /// fn main() {
+    ///     // SomeStuct is dropped immediately instead of at end of scope.
+    ///     let _ = SomeStruct;
+    /// }
+    /// ```
+    /// ### Explanation
+    ///
+    /// Statements which assign an expression to an underscore causes the
+    /// expression to immediately drop. Usually, it's better to explicitly handle
+    /// the `must_use` expression.
+    pub LET_UNDERSCORE_MUST_USE,
+    Allow,
+    "non-binding let on a expression marked `must_use`"
+}
+
+declare_lint_pass!(LetUnderscore => [LET_UNDERSCORE_DROP, LET_UNDERSCORE_LOCK, LET_UNDERSCORE_MUST_USE]);
+
+const SYNC_GUARD_PATHS: [&[&str]; 5] = [
+    &["std", "sync", "mutex", "MutexGuard"],
+    &["std", "sync", "rwlock", "RwLockReadGuard"],
+    &["std", "sync", "rwlock", "RwLockWriteGuard"],
+    &["parking_lot", "raw_mutex", "RawMutex"],
+    &["parking_lot", "raw_rwlock", "RawRwLock"],
+];
+
+impl<'tcx> LateLintPass<'tcx> for LetUnderscore {
+    fn check_local(&mut self, cx: &LateContext<'_>, local: &hir::Local<'_>) {
+        if !matches!(local.pat.kind, hir::PatKind::Wild) {
+            return;
+        }
+        if let Some(init) = local.init {
+            let init_ty = cx.typeck_results().expr_ty(init);
+            let needs_drop = init_ty.needs_drop(cx.tcx, cx.param_env);
```

Comment:
Can we exit early if `needs_drop` is false?  The later computations seem a bit heavy.

Reply:
Bailing out early would mean that types which are `must_use` but dont impl `Drop` won't get checked. Doing this would probably make `let_underscore_must_use` not very useful. (I am okay with just only having `let_underscore_lock` as lint, given that Clippy sets `let_underscore_must_use` and `let_underscore_drop` as allow-by-default)

Reply:
I decided to just remove `let_underscore_must_use` and implemented the early bailout instead.